### PR TITLE
Improve cross compiling support

### LINKED ISSUE #75: Suggestion: Windows CI
Partially related to #28.

Would you be receptive to a PR adding Windows CI? I'm in the process of submitting a PR to FlatBuffers and I'm seeing what looks like invalid generated cmake commands [on their AppVeyor CI jobs](https://ci.appveyor.com/project/gwvo/flatbuffers/builds/23473434):

```
--- stdout
running: "cmake" "C:\\projects\\flatbuffers\\rust\\flatc\\..\\.." "-Thost=x64" "-DCMAKE_GENERATOR_PLATFORM=x64" "-G" "Visual Studio 10 2010" "-DCMAKE_BUILD_TYPE=Release" "-DCMAKE_INSTALL_BINDIR=bin" "-DFLATBUFFERS_BUILD_TESTS=OFF" "-DFLATBUFFERS_BUILD_FLATLIB=OFF" "-DFLATBUFFERS_BUILD_FLATHASH=OFF" "-DCMAKE_INSTALL_PREFIX=C:\\projects\\flatbuffers\\tests\\rust_usage_test\\target\\debug\\build\\flatc-a64531bc5efc6cfc\\out" "-DCMAKE_C_FLAGS= /nologo /MD" "-DCMAKE_CXX_FLAGS= /nologo /MD"
-- Found Windows SDK v7.1: C:\Program Files\Microsoft SDKs\Windows\v7.1\
-- Configuring incomplete, errors occurred!
See also "C:/projects/flatbuffers/tests/rust_usage_test/target/debug/build/flatc-a64531bc5efc6cfc/out/build/CMakeFiles/CMakeOutput.log".
--- stderr
CMake Error at CMakeLists.txt:6 (project):
  Generator
    Visual Studio 10 2010
  given toolset specification
    host=x64
  that contains invalid field 'host=x64'.
```

Seems to me like it should generate `cmake -G "Visual Studio 10 2010 Win64"` *without* `-Thost=x64` as pointed out in the prior issue.

### LINKED ISSUE #80: Improve cross-compiling support
I ran into a cross-compiling issue where `cmake` used the host's `CMAKE_SYSTEM_PROCESSOR` instead of the target's and incorrectly evaluated a condition in the `CMakeLists.txt` that led to some unsupported compiler flags being used.

After thinking for a bit about how to deal with this issue in general, I can see a couple of solutions:
1. Use a [`CMAKE_TOOLCHAIN_FILE`](https://cmake.org/cmake/help/latest/variable/CMAKE_TOOLCHAIN_FILE.html) (#10) as suggested by the [CMake CrossCompiling wiki page](https://gitlab.kitware.com/cmake/community/wikis/doc/cmake/CrossCompiling)
2. Set the `CMAKE_SYSTEM_NAME` and `CMAKE_SYSTEM_PROCESSOR` in the `build.rs` for each target triplet that I need

The first seems too involved if I only need the second, but the second seems like it should be something not specific to my crate since it only depends directly on the `$TARGET`.

(Also, let me add as a warning that I am by no means very knowledgeable about CMake, so there may be other reasonable ways to do this.)

I see that there is some support for cross-compiling to certain targets in `cmake-rs` already (e.g. a9f74a4bf488c132876c9976c98b44cdc1a0603c, #35, #45), but it seems to be done on a case-by-case basis. Plus, there appears to be a large amount of target-specific logic 

So, this has led me to wonder if the `cmake-rs` cross-compiling/target-specific story could be improved and extended to support more targets in a relatively clean way. I'm posting this issue basically to see if there's interest and to discuss ideas. Here are some thoughts on different approaches:

1. Leave `cmake-rs` mostly untouched and continue to add special cases for various targets.

2. Create a sort of condition matrix to support combinations in a regular way. For example, you might start with `host == target` and `host != target` and branch within those into different supported targets.

3. Remove all or most of the target-specific logic and put it into `CMAKE_TOOLCHAIN_FILE`s, one for each of the different cases supported.

I suppose the latter two options have in common a mapping from Rust target triples to (at least) `CMAKE_SYSTEM_NAME`/`CMAKE_SYSTEM_PROCESSOR` pairs, though the Rust one may be more explicit. (I'm certainly not the first person to think of this: https://github.com/rust-lang/rust/issues/50220#issuecomment-433133136)

There are a few other improvements (at least, I think they are) that might also be considered (now or in some distant future):

* Use `$CARGO_CFG_TARGET_OS`, `$CARGO_CFG_TARGET_ARCH`, etc. instead of `$TARGET`
* CI with [`cross`](https://github.com/rust-embedded/cross), perhaps in the style of [`trust`](https://github.com/japaric/trust)
* iOS support (e.g. [`grpc-sys`](https://github.com/pingcap/grpc-rs/blob/fec97d193a2c25c43dc61db6e1d8716e17ee9dcf/grpc-sys/build.rs#L86-L113))

Anyway, these are just some initial thoughts on the matter. Feedback would be appreciated!

### LINKED ISSUE #151: Don't set CMAKE_SYSTEM_NAME=SunOS (and probably others)
This has been seen before, notably #138, but setting CMAKE_SYSTEM_NAME causes cmake to be in a cross-compiling configuration even when it's building native.  This causes rust-lang/rust#96054.  Only set CMAKE_SYSTEM_NAME when explicitly cross-compiling?

### PR DESCRIPTION
**Pull request by messense**
This PR changes `cmake-rs` to fill out `CMAKE_SYSTEM_NAME` and `CMAKE_SYSTEM_PROCESSOR` based `CARGO_CFG_TARGET_OS` and `CARGO_CFG_TARGET_ARCH` when cross compiling while `CMAKE_TOOLCHAIN_FILE` and `CMAKE_SYSTEM_NAME` isn't set by user.

It's tested on CI using [`cross`](https://github.com/cross-rs/cross).

Closes #75
Closes #80 
Closes #151

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -443,6 +443,59 @@ impl Config {
...
             if let Some(s) = self.getenv_target_os("CMAKE_TOOLCHAIN_FILE") {
                 self.define("CMAKE_TOOLCHAIN_FILE", s);
+            } else {
+                if target.contains("redox") {
+                    if !self.defined("CMAKE_SYSTEM_NAME") {
+                        self.define("CMAKE_SYSTEM_NAME", "Generic");
+                    }
+                } else if target.contains("solaris") {
+                    if !self.defined("CMAKE_SYSTEM_NAME") {
+                        self.define("CMAKE_SYSTEM_NAME", "SunOS");
```
Comment (messense) [inline src/lib.rs:453]: We can also address https://github.com/rust-lang/cmake-rs/issues/151 by removing this if it's acceptable.

Comment (gco) [reply src/lib.rs:453]: Yes, please!

**Review by thomcc — APPROVED** [review]
This looks great. Thanks for improving the CI too.

**Comment by jrose-signal** [comment]
I'm afraid this broke no-toolchain cross-compilation using Visual Studio, from an x86_64 Windows host to aarch64-pc-windows-msvc. Things are a bit clunky there in general (the existing support goes through [platforms](https://cmake.org/cmake/help/latest/generator/Visual%20Studio%2017%202022.html) and does not set `CMAKE_SYSTEM_PROCESSOR`), but it was working in 1.48.0 (and I'm afraid we hadn't been testing with new crate versions, so we didn't discover this sooner).

**Comment by jrose-signal** [comment]
Workaround: set `CMAKE_SYSTEM_NAME` to an empty string, which counts as defined for this crate, but unset for CMake itself.

**Comment by jrose-signal** [comment]
Turns out this *also* broke cross-compiling to Mac Catalyst targets (`*-apple-ios-macabi`), because CMake doesn't like having `CMAKE_SYSTEM_NAME` set to "iOS" while compiling against a macOS sysroot. That should probably eventually be fixed in CMake itself, but for now setting `CMAKE_SYSTEM_NAME` back to "Darwin" does work, at least for simple libraries. I don't know if you want to put that in this crate, though.

**Comment by thomcc** [comment]
Huh, this seems pretty bad!

**Pull request merged**